File: src/ingest.ts
#!/usr/bin/env node
/**
 * Ingest Arabic accent knowledge from PDFs and JSON into markdown files.
 *
 * Sources:
 *   - knowledge/*.pdf → extract text, output to data/<name>.md
 *   - knowledge/contributions/<id>/ → accent.json + PDFs or content.md
 *
 * Usage:
 *   npx arabic-accents-mcp ingest
 *   node dist/ingest.js
 */

import { readFileSync, writeFileSync, existsSync, mkdirSync, readdirSync, statSync } from "fs";
import { join, dirname, basename } from "path";
import { fileURLToPath } from "url";
// @ts-ignore - pdf-parse doesn't have great types
import pdf from "pdf-parse";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Directories
const ROOT_DIR = join(__dirname, "..");
const KNOWLEDGE_DIR = process.argv[2] || join(ROOT_DIR, "knowledge");
const DATA_DIR = join(ROOT_DIR, "data");
const CONTRIBUTIONS_DIR = join(KNOWLEDGE_DIR, "contributions");

interface AccentMeta {
  name: string;
  description: string;
}

interface Registry {
  [key: string]: AccentMeta;
}

// Extract text from PDF
async function extractPdfText(pdfPath: string): Promise<string> {
  try {
    const buffer = readFileSync(pdfPath);
    const data = await pdf(buffer);
    return data.text || "";
  } catch (error) {
    console.error(`  [ERROR] Failed to read ${basename(pdfPath)}: ${error}`);
    return "";
  }
}

// Convert PDF text to markdown
function pdfToMarkdown(text: string, name: string): string {
  return [
    `# ${name} Arabic\n`,
    `Content extracted from ${name} reference material.\n`,
    "---\n",
    text,
  ].join("\n");
}

// Process a contribution folder
async function processContribution(
  contribDir: string
): Promise<{ accentId: string; content: string; meta: AccentMeta } | null> {
  const accentJsonPath = join(contribDir, "accent.json");
  const contentMdPath = join(contribDir, "content.md");
  const dirName = basename(contribDir);

  if (!existsSync(accentJsonPath)) {
    console.log(`  [SKIP] ${dirName}: no accent.json`);
    return null;
  }

  let meta: AccentMeta;
  try {
    meta = JSON.parse(readFileSync(accentJsonPath, "utf-8"));
  } catch (error) {
    console.error(`  [ERROR] ${dirName}/accent.json: ${error}`);
    return null;
  }

  const name = meta.name || dirName;
  const description = meta.description || "";
  const accentId = dirName.toLowerCase();

  // Option 1: content.md exists
  if (existsSync(contentMdPath)) {
    const content = readFileSync(contentMdPath, "utf-8");
    return { accentId, content, meta: { name, description } };
  }

  // Option 2: PDFs in folder
  const pdfs = readdirSync(contribDir).filter(f => f.endsWith(".pdf"));
  if (pdfs.length > 0) {
    const textParts: string[] = [];
    for (const pdfFile of pdfs) {
      console.log(`    Extracting ${pdfFile}...`);
      const text = await extractPdfText(join(contribDir, pdfFile));
      if (text) {
        textParts.push(text);
      }
    }
    if (textParts.length > 0) {
      const combined = textParts.join("\n\n---\n\n");
      const md = `# ${name}\n\n${description}\n\n---\n\n${combined}`;
      return { accentId, content: md, meta: { name, description } };
    }
  }

  console.log(`  [SKIP] ${dirName}: no content.md or PDFs`);
  return null;
}

async function main() {
  console.log("Arabic Accents Ingestion Script\n");

  // Ensure data directory exists
  if (!existsSync(DATA_DIR)) {
    mkdirSync(DATA_DIR, { recursive: true });
  }

  const accents: Map<string, string> = new Map();
  const registry: Registry = {};

  // 1. Process PDFs in knowledge/
  console.log(`Processing PDFs in ${KNOWLEDGE_DIR}...`);
  if (existsSync(KNOWLEDGE_DIR)) {
    const files = readdirSync(KNOWLEDGE_DIR).filter(f => f.endsWith(".pdf"));
    for (const file of files) {
      console.log(`  ${file}`);
      const text = await extractPdfText(join(KNOWLEDGE_DIR, file));
      if (text) {
        const accentId = basename(file, ".pdf")
          .toLowerCase()
          .replace(/\s+/g, "-")
          .replace(/_/g, "-");
        const name = basename(file, ".pdf")
          .replace(/-/g, " ")
          .replace(/_/g, " ")
          .replace(/\b\w/g, c => c.toUpperCase());

        accents.set(accentId, pdfToMarkdown(text, name));
        registry[accentId] = {
          name: `${name} Arabic`,
          description: `Content from ${file}`,
        };
      }
    }
  }

  // 2. Process contributions
  if (existsSync(CONTRIBUTIONS_DIR)) {
    console.log(`\nProcessing contributions in ${CONTRIBUTIONS_DIR}...`);
    const dirs = readdirSync(CONTRIBUTIONS_DIR).filter(d => {
      const fullPath = join(CONTRIBUTIONS_DIR, d);
      return statSync(fullPath).isDirectory() && !d.startsWith(".");
    });

    for (const dir of dirs) {
      console.log(`  ${dir}/`);
      const result = await processContribution(join(CONTRIBUTIONS_DIR, dir));
      if (result) {
        const existing = accents.get(result.accentId);
        if (existing) {
          accents.set(result.accentId, existing + "\n\n---\n\n" + result.content);
        } else {
          accents.set(result.accentId, result.content);
          registry[result.accentId] = result.meta;
        }
      }
    }
  }

  // 3. Write output files
  console.log(`\nWriting ${accents.size} accent file(s) to ${DATA_DIR}...`);
  for (const [accentId, content] of accents) {
    const outPath = join(DATA_DIR, `${accentId}.md`);
    writeFileSync(outPath, content, "utf-8");
    console.log(`  ${accentId}.md`);
  }

  // 4. Write registry
  const registryPath = join(DATA_DIR, "accents.json");
  writeFileSync(registryPath, JSON.stringify(registry, null, 2), "utf-8");
  console.log(`  accents.json (registry)`);

  console.log(`\nDone. ${accents.size} accent(s) ready.`);
}

main().catch(console.error);

File: ingest.py
#!/usr/bin/env python3
"""
Ingest Arabic accent knowledge from PDFs and JSON into markdown files.

Sources:
  - knowledge/*.pdf          → extract text, output to data/<name>.md
  - knowledge/contributions/<id>/ → accent.json + PDFs or content.md
  - data/dialects/*.json     → curated JSON vocabulary → data/<dialect>.md

Usage:
  python ingest.py
  python ingest.py --knowledge-dir path/to/pdfs
"""

import argparse
import json
import re
from pathlib import Path

try:
    from pypdf import PdfReader
except ImportError:
    PdfReader = None


def extract_pdf_text(pdf_path: Path) -> str:
    """Extract text from a PDF file."""
    if PdfReader is None:
        print(f"  [WARN] pypdf not installed, skipping {pdf_path.name}")
        return ""
    try:
        reader = PdfReader(pdf_path)
        text_parts = []
        for page in reader.pages:
            text = page.extract_text()
            if text:
                text_parts.append(text)
        return "\n\n".join(text_parts)
    except Exception as e:
        print(f"  [ERROR] Failed to read {pdf_path.name}: {e}")
        return ""


def json_to_markdown(data: dict) -> str:
    """Convert curated JSON vocabulary to markdown."""
    dialect = data.get("dialect", "Unknown")
    terms = data.get("terms", [])
    
    lines = [f"# {dialect} Arabic\n"]
    lines.append(f"Vocabulary and phrases for {dialect} dialect.\n")
    lines.append("## Vocabulary\n")
    
    for term in terms:
        word = term.get("word", "").strip()
        meaning = term.get("meaning", "").strip()
        context = term.get("context", "").strip()
        
        if not word:
            continue
        
        lines.append(f"### {word}\n")
        if meaning:
            # Clean up meaning (remove excessive newlines)
            meaning_clean = re.sub(r'\n{3,}', '\n\n', meaning)
            lines.append(f"**Meaning:** {meaning_clean}\n")
        if context and context != meaning:
            context_clean = re.sub(r'\n{3,}', '\n\n', context)
            lines.append(f"**Example:** {context_clean}\n")
        lines.append("")
    
    return "\n".join(lines)


def pdf_to_markdown(text: str, name: str) -> str:
    """Wrap extracted PDF text in markdown structure."""
    lines = [f"# {name} Arabic\n"]
    lines.append(f"Content extracted from {name} reference material.\n")
    lines.append("---\n")
    lines.append(text)
    return "\n".join(lines)


def process_contribution(contrib_dir: Path) -> tuple[str, str] | None:
    """Process a contribution folder (accent.json + PDFs or content.md)."""
    accent_json = contrib_dir / "accent.json"
    content_md = contrib_dir / "content.md"
    
    if not accent_json.exists():
        print(f"  [SKIP] {contrib_dir.name}: no accent.json")
        return None
    
    try:
        meta = json.loads(accent_json.read_text(encoding="utf-8"))
    except Exception as e:
        print(f"  [ERROR] {contrib_dir.name}/accent.json: {e}")
        return None
    
    name = meta.get("name", contrib_dir.name)
    description = meta.get("description", "")
    accent_id = contrib_dir.name.lower()
    
    # Option 1: content.md exists
    if content_md.exists():
        content = content_md.read_text(encoding="utf-8")
        return accent_id, content
    
    # Option 2: PDFs in folder
    pdfs = list(contrib_dir.glob("*.pdf"))
    if pdfs:
        text_parts = []
        for pdf in pdfs:
            print(f"    Extracting {pdf.name}...")
            text = extract_pdf_text(pdf)
            if text:
                text_parts.append(text)
        if text_parts:
            combined = "\n\n---\n\n".join(text_parts)
            md = f"# {name}\n\n{description}\n\n---\n\n{combined}"
            return accent_id, md
    
    print(f"  [SKIP] {contrib_dir.name}: no content.md or PDFs")
    return None


def main():
    parser = argparse.ArgumentParser(description="Ingest Arabic accent knowledge")
    parser.add_argument("--knowledge-dir", type=Path, default=Path("knowledge"),
                        help="Directory containing PDFs (default: knowledge/)")
    parser.add_argument("--data-dir", type=Path, default=Path("data"),
                        help="Output directory for markdown files (default: data/)")
    args = parser.parse_args()
    
    knowledge_dir = args.knowledge_dir.resolve()
    data_dir = args.data_dir.resolve()
    dialects_dir = data_dir / "dialects"
    contributions_dir = knowledge_dir / "contributions"
    
    data_dir.mkdir(exist_ok=True)
    
    accents: dict[str, str] = {}  # accent_id -> markdown content
    registry: dict[str, dict] = {}  # accent_id -> {name, description}
    
    # 1. Process PDFs in knowledge/
    print(f"Processing PDFs in {knowledge_dir}...")
    for pdf in knowledge_dir.glob("*.pdf"):
        print(f"  {pdf.name}")
        text = extract_pdf_text(pdf)
        if text:
            accent_id = pdf.stem.lower().replace(" ", "-").replace("_", "-")
            # Try to make a nicer name
            name = pdf.stem.replace("-", " ").replace("_", " ").title()
            accents[accent_id] = pdf_to_markdown(text, name)
            registry[accent_id] = {"name": f"{name} Arabic", "description": f"Content from {pdf.name}"}
    
    # 2. Process curated JSON in data/dialects/
    if dialects_dir.exists():
        print(f"Processing curated JSON in {dialects_dir}...")
        for json_file in dialects_dir.glob("*.json"):
            print(f"  {json_file.name}")
            try:
                data = json.loads(json_file.read_text(encoding="utf-8"))
                dialect = data.get("dialect", json_file.stem)
                accent_id = dialect.lower().replace(" ", "-")
                md = json_to_markdown(data)
                
                # Merge if accent already exists
                if accent_id in accents:
                    accents[accent_id] += f"\n\n---\n\n{md}"
                else:
                    accents[accent_id] = md
                    registry[accent_id] = {"name": f"{dialect} Arabic", "description": f"Vocabulary for {dialect} dialect"}
            except Exception as e:
                print(f"  [ERROR] {json_file.name}: {e}")
    
    # 3. Process contributions
    if contributions_dir.exists():
        print(f"Processing contributions in {contributions_dir}...")
        for contrib in contributions_dir.iterdir():
            if contrib.is_dir() and not contrib.name.startswith("."):
                print(f"  {contrib.name}/")
                result = process_contribution(contrib)
                if result:
                    accent_id, content = result
                    if accent_id in accents:
                        accents[accent_id] += f"\n\n---\n\n{content}"
                    else:
                        accents[accent_id] = content
                        # Try to get name from accent.json
                        try:
                            meta = json.loads((contrib / "accent.json").read_text(encoding="utf-8"))
                            registry[accent_id] = {"name": meta.get("name", accent_id), "description": meta.get("description", "")}
                        except:
                            registry[accent_id] = {"name": accent_id.title(), "description": ""}
    
    # 4. Write output files
    print(f"\nWriting {len(accents)} accent file(s) to {data_dir}...")
    for accent_id, content in accents.items():
        out_path = data_dir / f"{accent_id}.md"
        out_path.write_text(content, encoding="utf-8")
        print(f"  {out_path.name}")
    
    # 5. Write registry
    registry_path = data_dir / "accents.json"
    registry_path.write_text(json.dumps(registry, indent=2, ensure_ascii=False), encoding="utf-8")
    print(f"  {registry_path.name} (registry)")
    
    print(f"\nDone. {len(accents)} accent(s) ready.")


if __name__ == "__main__":
    main()
